### ISSUE TITLE
Is it possible to directly pass a PointCloud XYZ to small GICP for registration? Are there any corresponding examples?

### ISSUE THREAD

**Issue by JACKLiuDay** [koide3/small_gicp#98]
Hello, I noticed that in the examples in the build, the point clouds are passed in as Eigen::Vector4f format. However, in practical use, we often work with the original PointCloud XYZ format. How can we pass the point cloud for registration?

**Comment by JACKLiuDay** [comment]
auto target_tree = std::make_shared<small_gicp::KdTree<small_gicp::PointCloud>>(target, small_gicp::KdTreeBuilderOMP(num_threads)); // Create KdTree
    auto source_tree = std::make_shared<small_gicp::KdTree<small_gicp::PointCloud>>(source, small_gicp::KdTreeBuilderOMP(num_threads));

    small_gicp::estimate_covariances_omp(*target, *target_tree, num_neighbors, num_threads);                    // Estimate point covariances
    small_gicp::estimate_covariances_omp(*source, *source_tree, num_neighbors, num_threads);

Hello, I would like to know if covariance estimation must be performed before registration, or if it is possible to directly align the point cloud after downsampling without this step?

**Comment by koide3** [comment]
> However, in practical use, we often work with the original PointCloud XYZ format. How can we pass the point cloud for registration?

You can use `RegistrationPCL<pcl::PointXYZ, pcl::PointXYZ>` to work with PCL point cloud. 

https://github.com/koide3/small_gicp/blob/master/src/example/02_basic_registration_pcl.cpp

It is also possible to feed PCL point clouds (and any other point cloud types) to `Registration` template by only defining data access traits as in the following example.

https://github.com/koide3/small_gicp/blob/master/src/example/03_registration_template.cpp

> I would like to know if covariance estimation must be performed before registration, or if it is possible to directly align the point cloud after downsampling without this step?

You need to perform covariance estimation (if you use GICP) before registration in this way.

https://github.com/koide3/small_gicp/blob/master/src/example/01_basic_registration.cpp

**Comment by JACKLiuDay** [comment]
> > However, in practical use, we often work with the original PointCloud XYZ format. How can we pass the point cloud for registration?
> 
> You can use `RegistrationPCL<pcl::PointXYZ, pcl::PointXYZ>` to work with PCL point cloud.
> 
> https://github.com/koide3/small_gicp/blob/master/src/example/02_basic_registration_pcl.cpp
> 
> It is also possible to feed PCL point clouds (and any other point cloud types) to `Registration` template by only defining data access traits as in the following example.
> 
> https://github.com/koide3/small_gicp/blob/master/src/example/03_registration_template.cpp
> 
> > I would like to know if covariance estimation must be performed before registration, or if it is possible to directly align the point cloud after downsampling without this step?
> 
> You need to perform covariance estimation (if you use GICP) before registration in this way.
> 
> https://github.com/koide3/small_gicp/blob/master/src/example/01_basic_registration.cpp

Thank you very much for your reply. Initially, I thought that small GICP only supports input data in the form of Eigen::Vector4f, and that PCL PointCloudXYZ would need to be converted to Eigen::Vector4f before registration. Additionally, I would like to ask about the covariance calculation. If using VGICP, is it still necessary to compute the covariance beforehand? My understanding is that covariance calculation is only required when using the GICP algorithm.

**Comment by koide3** [comment]
Yes, VGICP also requires covariance computation because it computes voxel covs by aggregating point covs.

**Comment by JACKLiuDay** [comment]
> Yes, VGICP also requires covariance computation because it computes voxel covs by aggregating point covs.

Thank you very much for your reply. We will try using Small GICP in our project.
